src/flow/ui/FlowStackNavigator.tsx

```diff
@@ -1,55 +1,50 @@
-
-  // Every time the initial route name changes to a different string, reconstruct the entire thing.
-  // Why? The initial route name will only be set once(!), i.e. during the initial render.
-  // Subsequent changes won't alter the initial route.
-  return (
-    <FlowStack.Navigator>
-      {getFlowStackScreenRoutes(initialRouteName).map((route) => {
-        const screen = FLOW_STACK_SCREENS[route];
-        return (
-          <FlowStack.Screen
-            key={route}
-            name={route} // e.g. "Step1"
-            component={screen} // e.g. Step1Screen
-            options={{
-              headerShown: !routeIsFlowStackPreInitialRouteName(route),
-              // Force the animation behavior for initial route such that
-              // - going back to the pre-initial route is pop
-              // - going forward from initial route is push
-              // - everything else is the navigation framework default, whatever that value is
-              //
-              // The point of this forced animation is to make the transition phase
-              // between the Android fragments and the React screens look consistent and natural.
-              animationTypeForReplace:
-                routeIsFlowStackInitialRouteName(route) || routeIsFlowStackPreInitialRouteName(route)
-                  ? prevInitialRouteName == null && initialRouteName != null
-                    ? 'push'
-                    : 'pop'
-                  : undefined,
-            }}
-          />
-        );
-      })}
-    </FlowStack.Navigator>
-  );
-};
+const FlowStackNavigatorBase = createSwitchNavigator(
+  {
+    ...FLOW_STACK_ENTRY_POINT_SCREENS,
+    [FLOW_STACK_STACK_ROUTES.Login]: createStackNavigator(FLOW_STACKS.LoginEntry),
+    [FLOW_STACK_STACK_ROUTES.Register]: createStackNavigator(FLOW_STACKS.RegisterEntry),
+  },
+  {
+    initialRouteName: LoadingScreen.ROUTE,
+    backBehavior: 'none',
+  },
+);
+
+const FlowStackNavigator = ((props: NavigationInjectedProps) => {
+  const { initialRouteName } = useInitialRouteNameStateContext(InitialRouteNameStateContext);
+  const { navigation } = props;
+
+  useEffect(() => {
+    // Jump into the stack which has the entry point we want, not the route inside that stack.
+    navigation.navigate(initialRouteName ? initialRouteName + 'Entry' : LoadingScreen.ROUTE);
+  }, [initialRouteName]);
+
+  return <FlowStackNavigatorBase {...props} />;
+}) as typeof FlowStackNavigatorBase;
+
+// This is intentional.
+// createNavigator and others that wrap over it (createStackNavigator, createSwitchNavigator) make
+// use of class static members for the express purpose of routing through each screen.
+// Without hoisting these class static members, routing will refuse to work.
+// https://github.com/react-navigation/react-navigation/blob/4.x/packages/core/src/navigators/createNavigator.js#L8-L11
+hoistNonReactStatics(FlowStackNavigator, FlowStackNavigatorBase);
```

Comment:
# Authentication Flow technique

This is the most important part. Like in https://github.com/csantarin/android-startup-perf-app/pull/37#discussion_r1027501929, the root-level navigator here is never unmounted. However, there's a fine difference here:

| version | method | conservative number of possible stacks loaded |
| --- | --- | --- |
| 6.x.x | conditional rendering at run time | always one stack only |
| 4.x.x. | static determination at initial run time of all possibilities | one stack per entry point |

In 4.x.x., conditional rendering at run time is not possible short of possibly implementing the stack entry points using some sort of object (React ref?). It is likely impractical to have to reinvent the wheel in such a scenario.

## References

https://reactnavigation.org/docs/4.x/auth-flow